### ISSUE TITLE
Add read_only parameter to Pydantic Field

### ISSUE THREAD

**Issue by Davixk** [pydantic/pydantic#9917]
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this feature is missing

## Use Case

I'm working on a project where I have a base model that subclasses both \`pydantic.BaseModel\` and \`abc.ABC\`. I need to define fields in this base model with the following properties:

1. Read-only from an external perspective
2. Must be set by subclasses (not set in the base model)
3. Flexible implementation in subclasses (can be a simple attribute, property, or computed field)

Here's a simplified example of what I'm trying to achieve:

```python
from abc import ABC
from pydantic import BaseModel, Field

class Foo(BaseModel, ABC):
    readonly_field: str = Field(...)  # Needs to be read-only and implemented by subclasses

class Bar(Foo):
    readonly_field: str = "simple value"  # Works fine

class Baz(Foo):
    @property
    def readonly_field(self) -> str:  # Should also work
        return "computed value"

# Usage
bar = Bar()
print(bar.readonly_field)  # Output: simple value
bar.readonly_field = "new value"  # Should raise an error, but doesn't

baz = Baz()
print(baz.readonly_field)  # Output: computed value
```

## Current Limitations

The existing Pydantic features don't fully address this use case:

1. `Field(init=False)`: This prevents the field from being set during initialization, but it doesn't make it read-only after creation.

2. `frozen=True` (model-level or field-level): This makes the entire model or field immutable, which is too restrictive. It also doesn't allow for computed properties in subclasses.

3. `@property` in the base class: This forces all subclasses to implement it as a property, reducing flexibility and cluttering the code with boilerplate for simple values.

### Example with Property (Less Flexible and Clean)
```python
class Foo(BaseModel, ABC):
    @property
    def readonly_field(self) -> str:
        raise NotImplementedError

class Bar(Foo):
    @property
    def readonly_field(self) -> str:
        return "simple value"

class Baz(Foo):
    @property
    def readonly_field(self) -> str:
        return "computed value"
```

### Proposed Solution (More Flexible and Clean)
```python
class Foo(BaseModel, ABC):
    readonly_field: str = Field(..., read_only=True)

class Bar(Foo):
    readonly_field: str = "simple value"  # Clean and simple

class Baz(Foo):
    @property
    def readonly_field(self) -> str:  # Flexibility for computed values
        return "computed value"
```

## Proposed Solution

Add a `read_only` parameter to the `Field` function (or whatever other name):

```python
class Foo(BaseModel, ABC):
    readonly_field: str = Field(..., read_only=True)
```

This would:
1. Allow the field to be set once during object creation or in a subclass.
2. Prevent the field from being modified after initialization.
3. Work with different implementation styles in subclasses (attribute, property, computed field).
5. Maintain type hinting and other Pydantic features.

## Benefits

Adding a `read_only` parameter would:
1. Provide more granular control over field mutability.
2. Allow for flexible implementation in subclasses while maintaining read-only behavior.
3. Improve type safety and reduce the need for custom descriptors or workarounds.
5. Align with common patterns in other ORMs and data modeling libraries.
6. Maintain the simplicity and clarity that Pydantic users expect.

Would love to hear thoughts on this request. Thanks for considering!

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [ ] [Data validation/parsing](https://docs.pydantic.dev/concepts/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/concepts/serialization/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/concepts/json_schema/)
- [X] [Dataclasses](https://docs.pydantic.dev/concepts/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/concepts/config/)
- [X] [Field Types](https://docs.pydantic.dev/api/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/concepts/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/concepts/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/concepts/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by Viicos** [comment]
I see two ways to do so:

- Using `ClassVar`:

```python
from abc import ABC
from typing import Any, Callable, ClassVar

from pydantic_core.core_schema import SerializationInfo
from typing_extensions import Self

from pydantic import BaseModel, model_serializer


class Foo(BaseModel, ABC):
    readonly_field: ClassVar[str]

    @model_serializer(mode='wrap')
    def serialize(self, handler: Callable[[Self], dict[str, Any]], info: SerializationInfo):
        # If you want `readonly_field` to be included during serialization
        data = handler(self)
        data['readonly_field'] = self.readonly_field
        return data


class Bar(Foo):
    readonly_field = 'test'


print(Bar().model_dump_json())
#> {"readonly_field":"test"}
```

- Using an abstract property (combined with `@computed_field`):

```python
from abc import ABC, abstractmethod

from pydantic import BaseModel, computed_field


class Foo(BaseModel, ABC):
    @property
    @abstractmethod
    def readonly_field(self) -> str:
        pass


class Bar(Foo):
    @computed_field
    @property
    def readonly_field(self) -> str:
        return 'test'


print(Bar().model_dump_json())
#> {"readonly_field":"test"}
```

Regarding:

> Flexible implementation in subclasses (can be a simple attribute, property, or computed field)

Seems like such flexibility isn't really possible. From a typing perspective, you also won't be able to do so as well (e.g. if `readonly_field` is defined as an abstract property it can't be defined as a `ClassVar` attribute in subclasses and vice-versa).

**Comment by Davixk** [comment]
Thanks for the reply @Viicos !

Unfortunately, neither of these solutions really accomplish what I'm requesting.

In my use case, I wouldn't really want the field to be a class variable, as if it's used as a property in a subclass I'd want it to to change at an instance level and not the class level. So essentially it being a class property would be a different attribute from what I'm asking.

As for simply using it as a property (and computed_field) both in the super and subclass - as I mentioned with examples above - it wouldn't provide the cleanliness and flexibility I'm proposing, although it would work.

With that said, I understand the level of flexibility I'm requesting probably isn't possible, but I still thought it would be worth putting in a feature request.

**Comment by Viicos** [comment]
Ah I see, so you mean a `ClassVar` isn't suitable because you may want to have this field depend on others? (or at least vary at between instances in some way)

**Comment by Davixk** [comment]
Exactly. I wouln't want it necessarily tied to the class, as it could change on an instance basis (wouldn't have to, but could).

**Comment by Viicos** [comment]
https://discuss.python.org/t/67359/ might be a good way to support this.

**Comment by Viicos** [comment]
> If and when this eventually gets implemented in a future version of Python, it would be great if it could be supported in Pydantic Fields via the `read_only` parameter.

I don't think a new `read_only` parameter will do. Instead, you will probably annotate your fields as `f: ReadOnly[...]` to gain type checker support.

**Comment by chimeziriobioha** [comment]
Excellent feature request, and great discussion.

**Comment by Davixk** [comment]
Quick update:

The PEP has shifted to PEP767. Here's a [link to the discourse](https://discuss.python.org/t/pep-767-annotating-read-only-attributes/73408), and here's a [link to the draft PEP](https://peps.python.org/pep-0767/).

I'll update this issue thread if and when the PEP is approved.
